### ISSUE TITLE
[FEATURE] Automatic Prompt Optimization for LangChain4j

### ISSUE THREAD

**Issue by omarmahamid** [langchain4j/langchain4j#4267]
Idea derived from 

https://github.com/stanfordnlp/dspy

# Problem Definition

## The Problem We're Solving
Current State: Manual Prompt Engineering

```mermaid
flowchart LR
    subgraph Current["Current Painful Process"]
        A[Developer] -->|writes| B[Initial Prompt]
        B -->|tests| C[Try on Examples]
        C -->|fails| D[Tweak Prompt]
        D -->|tests again| C
        C -->|seems ok| E[Deploy]
        E -->|users complain| F[Edge Cases Found]
        F -->|back to| D
    end
    
    style Current fill:#ffcccc
```

<section id="markdown-section-ab714f8f-e00d-403c-975a-74dfe339f3ee-7" class="markdown-section chat-fade-in" data-markdown-raw="

### Pain Points" data-section-index="7" style="border-radius: 4px; line-height: 19.5px; margin: 6px 0px; position: relative; scroll-margin-bottom: 40px; scroll-margin-top: 40px; color: rgba(228, 228, 228, 0.92); font-family: -apple-system, &quot;system-ui&quot;, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(20, 20, 20); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><h3 style="animation: 0.25s ease 0s 1 normal none running fade-in; font-weight: 600 !important; font-size: 1.15em; line-height: 1.25; margin-bottom: 8px; margin-top: 18px;"><span style="animation: 0.25s ease 0s 1 normal none running fade-in;">Pain</span><span style="animation: 0.25s ease 0s 1 normal none running fade-in;"><span> </span>Points</span></h3></section>
Problem | Impact | Example
-- | -- | --
Trial & Error | Days spent tweaking prompts | "Add 'step by step'... no wait, remove it..."
No Validation | Unknown quality until production | "It worked on my 5 test cases!"
Not Reproducible | Different results each time | "It was working yesterday!"
Expert Knowledge Needed | Bottleneck on prompt engineers | "Only Sarah knows how to write these"
Overfitting | Works on test cases, fails in prod | "Why doesn't it handle X?"

Desired State: Automatic Optimization

```mermaid
flowchart LR
    subgraph Desired["DSPy4j Approach"]
        A[Developer] -->|defines| B[Input/Output Types]
        A -->|provides| C[Training Examples]
        A -->|specifies| D[Success Metric]
        
        B --> E[DSPy4j Optimizer]
        C --> E
        D --> E
        
        E -->|generates| F[Optimized Prompt]
        F -->|validated| G["94% Accuracy ✓"]
        G --> H[Deploy with Confidence]
    end
    
    style Desired fill:#ccffcc
```

Concrete Problem Statement
> Given:
> - A task definition (input types → output types)
> - Training examples with expected outputs
> - A metric to evaluate quality
>
> Produce:
> - An optimized prompt that maximizes the metric
> - Validated performance score on held-out test set

**Comment by dliubarskyi** [comment]
@omarmahamid nice idea! Are you willing to implement it?

**Comment by omarmahamid** [comment]
@dliubarskyi yes sure

**Comment by dliubarskyi** [comment]
Great! Fel free to open a PR against a https://github.com/langchain4j/langchain4j-community repository and tag me there (@dliubarskyi)

**Comment by omarmahamid** [comment]
@dliubarskyi 

You don't want it to be here as new module maybe ?

**Comment by dliubarskyi** [comment]
@omarmahamid this features is quite big and, unfortunately, I don't have a capacity for reviewing and maintaining it in the near future. It will still be documented in the common documentation and available for all users.

**Comment by omarmahamid** [comment]
Got it

Will PR it to the community

**Comment by stefanofornari** [comment]
is there a PR for this?

**Comment by omarmahamid** [comment]
@stefanofornari not yet

In the oven will be pushed to community repo
